specs/capella/beacon-chain.md

```diff
@@ -0,0 +1,299 @@
+        index=state.withdrawal_index,
+        address=state.validators[index].withdrawal_credentials[12:],
+        amount=amount,
+    )
+    state.withdrawal_index = WithdrawalIndex(state.withdrawal_index + 1)
+    state.withdrawals_queue.append(withdrawal)
+```
+
+### Predicates
+
+#### `is_fully_withdrawable_validator`
+
+```python
+def is_fully_withdrawable_validator(validator: Validator, epoch: Epoch) -> bool:
+    """
+    Check if ``validator`` is fully withdrawable.
+    """
+    is_eth1_withdrawal_prefix = validator.withdrawal_credentials[0:1] == ETH1_ADDRESS_WITHDRAWAL_PREFIX
+    return is_eth1_withdrawal_prefix and validator.withdrawable_epoch <= epoch < validator.withdrawn_epoch
+```
+
+## Beacon chain state transition function
+
+### Epoch processing
+
+```python
+def process_epoch(state: BeaconState) -> None:
+    process_justification_and_finalization(state)
+    process_inactivity_updates(state)
+    process_rewards_and_penalties(state)
+    process_registry_updates(state)
+    process_slashings(state)
+    process_eth1_data_reset(state)
+    process_effective_balance_updates(state)
+    process_slashings_reset(state)
+    process_randao_mixes_reset(state)
+    process_historical_roots_update(state)
+    process_participation_flag_updates(state)
+    process_sync_committee_updates(state)
+    process_full_withdrawals(state)  # [New in Capella]
+```
+
+#### Withdrawals
+
+*Note*: The function `process_full_withdrawals` is new.
+
+```python
+def process_full_withdrawals(state: BeaconState) -> None:
+    current_epoch = get_current_epoch(state)
+    for index, validator in enumerate(state.validators):
+        if is_fully_withdrawable_validator(validator, current_epoch):
+            # TODO, consider the zero-balance case
+            withdraw(state, ValidatorIndex(index), state.balances[index])
+            validator.withdrawn_epoch = current_epoch
+```
+
+### Block processing
+
+```python
+def process_block(state: BeaconState, block: BeaconBlock) -> None:
+    process_block_header(state, block)
+    if is_execution_enabled(state, block.body):
+        process_withdrawals(state, block.body.execution_payload)  # [New in Capella]
+        process_execution_payload(state, block.body.execution_payload, EXECUTION_ENGINE)  # [Modified in Capella]
+    process_randao(state, block.body)
+    process_eth1_data(state, block.body)
+    process_operations(state, block.body)
+    process_sync_aggregate(state, block.body.sync_aggregate)
+```
+
+#### New `process_withdrawals`
+
+```python
+def process_withdrawals(state: BeaconState, payload: ExecutionPayload) -> None:
+    num_withdrawals = min(MAX_WITHDRAWALS_PER_PAYLOAD, len(state.withdrawals_queue))
+    dequeued_withdrawals = state.withdrawals_queue[:num_withdrawals]
+
+    assert len(dequeued_withdrawals) == len(payload.withdrawals)
+    for dequeued_receipt, withdrawal in zip(dequeued_withdrawals, payload.withdrawals):
+        assert dequeued_receipt == withdrawal
```

Comment:
we *must* check `payload.withdrawals` here -- this is how we know that the `payload` is well-formed with respect to the consensus `state`, and that this is indeed a valid block

otherwise a malicious proposer could stuff arbitrary withdrawals into the `payload` and feed them to the execution layer...